crates/sui-graphql-rpc/src/types/gas.rs

```diff
@@ -58,31 +61,51 @@ impl GasInput {
         &self,
         ctx: &Context<'_>,
         first: Option<u64>,
-        after: Option<object::Cursor>,
+        after: Option<CGasPayment>,
         last: Option<u64>,
-        before: Option<object::Cursor>,
+        before: Option<CGasPayment>,
     ) -> Result<Connection<String, Object>> {
         // A possible user error during dry run or execution would be to supply a gas payment that
         // is not a Move object (i.e a package). Even though the transaction would fail to run, this
         // service will still attempt to present execution results. If the return type of this field
         // is a `MoveObject`, then GraphQL will fail on the top-level with an internal error.
         // Instead, we return an `Object` here, so that the rest of the `TransactionBlock` will
         // still be viewable.
-        let page = Page::from_params(ctx.data_unchecked(), first, after, last, before)?;
 
-        let filter = ObjectFilter {
-            object_keys: Some(self.payment_obj_keys.clone()),
-            ..Default::default()
+        let mut connection = Connection::new(false, false);
+        // Return empty connection if no payment objects
+        if self.payment_obj_keys.is_empty() {
+            return Ok(connection);
+        }
+
+        let page = Page::from_params(ctx.data_unchecked(), first, after, last, before)?;
```

Comment:
This should be the first operation still, so that we consistently return an error if the pagination parameters are wrong.